### DISCUSSION TITLE
Calling APIs defined within WASM modules from Native

### DISCUSSION THREAD

**Discussion by ghost** [bytecodealliance/wasm-micro-runtime#3696] [Q&A] [answered]
I want to be able to call functions defined in WASM modules from my native library.
1 way I found to do this is with embedding WAMR in the application and looking up the WASM function using string.
Reference for Embed WAMR way: https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/doc/embed_wamr.md

However is there a way to achieve above even when I am running my WASM module with iwasm? It would seem like iwasm would run in its own process and memory of the WASM module is not directly accessible to native application running in a different process.

Is there any documented way to achieve exporting WASM APIs to native with iwasm?

**Comment by loganek** [comment, marked as answer]
Just so I understand, you have an exported wasm function, something like:

```
(module
   (export "foo" (func $foo))
   ; .....
)
```

that you could potentially generate from C:
```c
#include <stdio.h>

void foo() {
  printf("HELLO\n");
}
```
and compile with 
```bash
/opt/wasi-sdk/bin/clang app.c -o app.wasm -Wl,--export=foo
```
and you'd like to run this particular function from `iwasm`? If so, there's a `--function` parameter that you could use:
```bash
./iwasm --function foo app.wasm
```
Not sure though if that's what you've asked for.

**Reply by ghost** [reply]
I also want to be able to call this exported function from my native. How would I build native with the function symbol since it would complain about implicit declaration of the symbol.

**Reply by ghost** [reply]
If I use --function argument, it will tell iwasm the function to be called when wasm module is executed. 
In your above example, this still doesn't give me the provision to invoke 'foo' in native.
